File: js/bench.js
const minimist = require('minimist');
const World = require("./world.js");

const args = minimist(process.argv.slice(2));

const width = args.width || 100;
const height = args.height || 100;
const liveCount = args['live-count'] || 100;
const stepCount = args['step-count'] || 1000;
// Not supporting seed like other implementations because JS doesn't natively suport it.

let w = new World(width, height);

for (let i = 0; i < liveCount; i++) {
  w.set(Math.floor(Math.random() * width), Math.floor(Math.random() * height), true);
}

const startTime = Date.now();
for (let i = 0; i < stepCount; i++) {
  w = w.step();
}
const endTime = Date.now();

console.log(`JS: ${(endTime - startTime) / 1000}`);

File: ruby/bench.rb
require 'optparse'
require_relative "world"

options = {}
OptionParser.new do |parser|
  parser.banner = "Usage: cgol.rb [options]"

  parser.on("--width INTEGER", Integer, "Width") do |value|
    options[:width] = value
  end
  parser.on("--height INTEGER", Integer, "Height") do |value|
    options[:height] = value
  end
  parser.on("--live-count INTEGER", Integer, "Live cell count to start") do |value|
    options[:live_count] = value
  end
  parser.on("--step-count INTEGER", Integer, "Number of steps to run") do |value|
    options[:step_count] = value
  end
  parser.on("--seed INTEGER", Integer, "Seed for RNG") do |value|
    options[:seed] = value
  end
end.parse!

width = options.fetch(:width, 100)
height = options.fetch(:height, 100)
live_count = options.fetch(:live_count, 100)
step_count = options.fetch(:step_count, 1_000)
seed = options.fetch(:seed, 0)
srand(seed)

w = World.new(width, height)

live_count.times do
  w.set(rand(w.width), rand(w.height), true)
end

start_time = Time.now
step_count.times do |n|
  w = w.step
end
end_time = Time.now

duration = end_time - start_time
puts "Ruby: #{duration}"
